backend/regalloc/regalloc_split.ml

```diff
@@ -5,93 +5,107 @@ open! Regalloc_split_utils
+   tree in a prefix manner. [bindings] is the part of the substitutions
+   propagating from the parent node in the tree. In effect, the way names are
+   distributed is quite similar to what happens in SSA. *)
+let rec compute_substitution_tree :
     State.t ->
-    Cfg.t ->
     Substitution.map ->
-    Cfg.basic_block ->
-    Substitution.t ->
+    Reg.t Reg.Map.t ->
+    Cfg_dominators.dominator_tree ->
     unit =
- fun state cfg substs start subst ->
-  if split_debug then log ~indent:1 "propagate_substitution";
-  let visited = ref Label.Set.empty in
-  let rec iter (block : Cfg.basic_block) : unit =
-    if not (Label.Set.mem block.start !visited)
-    then (
-      if split_debug then log ~indent:2 "iter %d" block.start;
-      visited := Label.Set.add block.start !visited;
-      if Label.Tbl.mem substs block.start
-      then fatal "block %d already has a substitution" block.start;
-      Label.Tbl.replace substs block.start subst;
-      match destruction_point_at_end block with
-      | Some Destruction_on_all_paths -> ()
-      | None | Some Destruction_only_on_exceptional_path ->
-        let successor_labels =
-          Cfg.successor_labels block ~normal:true ~exn:false
-        in
-        Label.Set.iter
-          (fun successor_label ->
-            let definitions =
-              Label.Map.mem successor_label
-                (State.definitions_at_beginning state)
-            in
-            let phi =
-              Label.Map.mem successor_label (State.phi_at_beginning state)
-            in
-            if not (definitions || phi)
-            then iter (Cfg.get_block_exn cfg successor_label))
-          successor_labels)
-    else if split_debug
-    then log ~indent:2 "block %d already visited" block.start
+ fun state substs bindings tree ->
+  let label = tree.label in
+  if split_debug then log ~indent:1 "compute_substitution_tree %d" label;
+  (* First, remove the phi and definitions from the bindings. *)
+  if split_debug then log ~indent:2 "removing from phis";
+  let bindings =
+    remove_from_bindings state label ~field:State.phi_at_beginning
+      ~extract:Fun.id bindings
   in
-  iter start
-
-(* Computes the substitutions for all blocks, by introducing new registers at
-   reload points and then propagating the substitution until a new one takes
-   over. *)
-let compute_substitutions : State.t -> Cfg_with_infos.t -> Substitution.map =
- fun state cfg_with_infos ->
-  if split_debug then log ~indent:0 "compute_substitutions";
-  let cfg = Cfg_with_infos.cfg cfg_with_infos in
-  let substs = Label.Tbl.create (Label.Tbl.length cfg.blocks) in
-  let compute_substitution_for_block label =
-    if split_debug then log ~indent:1 "visiting block %d" label;
+  if split_debug then log ~indent:2 "removing from definitions";
+  let bindings =
+    remove_from_bindings state label ~field:State.definitions_at_beginning
+      ~extract:Fun.id bindings
```

Comment:
In what situation does this line have any effect on bindings? I mean, bindings are a result of a reload in a preceeding block, is it possible to have a reload again, without a spill between them on the same path  (that would have killed the binding) and without a the current block havig a phi too? This is safe to keep, just trying to confirm my understanding of what this code is doing.

Reply:
Yes, it it possible to have two reloads in a row.
_i.e._ with neither a spill or a phi. Spills may
disappear since we delete dominated spills of
constant values. If that happens on straight-line
code, there could be no phi as well.

Reply:
ah, yes, the constant spills. Worth a comment.